docs: add caching examples to README

villelahdenvuo opened the pull request:
* Added bundler and npm caching to examples
* Added scheduling example using `workflow_dispatch`

jonabc commented on README.md:
```diff
@@ -82,28 +84,35 @@ permissions:
+      - uses: jonabc/setup-licensed@v1
+        with:
+          version: 3.x
+      #region Example for Node projects
+      - uses: actions/setup-node@v2
+        with:
+          node-version: 16
+          cache: npm # cache dependencies for faster subsequent runs.
+      - run: npm install # install your projects dependencies in local environment
+      #endregion
```
what does `#region` and `#endregion` provide vs a normal comment? just curious if there's some built in editor or actions functionality I wasn't aware of 😄 

Good idea to add setup-node 👍

villelahdenvuo replied:
Yeah, VS Code supports code folding with `#region`/`#endregion` comments.

villelahdenvuo replied:
@jonabc It doesn't automatically fold anything (by default), but if more examples are added later on, it's easier to  run the command "fold all" and unfold the relevant regions.

However, if you don't like them, I can just add a normal comment.

jonabc commented on README.md:
````diff
@@ -128,21 +137,51 @@ permissions:
+          command: "bundle exec licensed" # or bin/licensed when using binstubs
+```
+
+### Scheduling licensed workflow to run regularly using `workflow_dispatch` event.
````
this isn't the [workflow_dispatch](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#workflow_dispatch) event.  I'm not sure I see why it makes sense to use `benc-uk/workflow-dispatch` when the user could just run the licensed action workflow from the schedule in the same workflow:

```yml
on:
   # run on pushes to the default branch
  push:
    branches:
      - main
  # run on pull request events with changes to code
  pull_request:
    types:
      - opened
      - reopened
      - synchronize
  # run on demand
  workflow_dispatch:
  # run on a schedule
  schedule:
    - cron: '0 10 * * 2' # e.g. every 10:00 on Tuesdays.
```

That said, TBH I'm not sure that I'd want to include a schedule trigger as an example here because (as written) action run failures aren't propagated anywhere outside of action workflow logs.  If something was run on a schedule I'd expect that any failures would need to be reported to an issue or some other notification mechanism.

villelahdenvuo replied:
We tried using schedule directly, but that event doesn't have the required properties that the `push`/`workflow_dispatch` events do, hence why the use of manual `workflow_dispatch` is needed.

villelahdenvuo replied:
Perhaps that could be fixed by allowing the user to specify the missing options e.g. branch

jonabc replied:
> Perhaps that could be fixed by allowing the user to specify the missing options e.g. branch

🚀  great suggestion!  I also ran some tests on what other contextual information is available when using `schedule` as the event trigger and it looks like `context.ref` is set to `refs/heads/<default branch name>`.

Between allowing users to manually set the branch and picking up the branch from `context.ref` for a broader number of use cases it sounds like this would drop the need for `benc-uk/workflow-dispatch`.  Is that right?

jonabc replied:
> But if the branch is the only information it uses from the event it should work the way you described.

The action also uses `context.repo`, but that gets filled from the `GITHUB_REPOSITORY` environment variable which looks like it's available for all of the tested `push`, `pull_request`, `workflow_dispatch`, `schedule` event triggers.

villelahdenvuo replied:
Opened https://github.com/jonabc/licensed-ci/issues/133

jonabc reviewed (commented):
🚀  thanks for the updates.  a bit of feedback on the changes, the only one I think is significant is around the scheduled workflow which I'm not sure is a great pattern as written.

the concern there is that it could lead to problems if someone copies/pastes the workflow verbatim to their repo and ends up with compliance/legal issues sometime in the future because they missed that failures were occurring.

villelahdenvuo commented on README.md:
````diff
@@ -82,28 +84,35 @@ permissions:
 ```
 
-Basic usage installing licensed gem using bundler + Gemfile
+### <a name="usage-bundler"></a>Basic usage installing licensed gem using bundler + Gemfile
````
That is true, but they are quite long in this case, that's why I added a custom one. But I can remove it if you prefer. Also the custom anchor is more durable if someone updates the title.

villelahdenvuo replied:
@jonabc Would you be okay with moving the Bundler example to be the first example? The reason being that it includes automatic caching unlike `setup-licensed`, which I noticed takes quite a while to install on every run.

jonabc replied:
Sure, I don't really have any preference on order here 👍

villelahdenvuo commented:
> the concern there is that it could lead to problems if someone copies/pastes the workflow verbatim to their repo and ends up with compliance/legal issues sometime in the future because they missed that failures were occurring.

I understand, perhaps it's better to add a note about the caveat. We have solved this issue by automatically creating an issue if the check fails and running it before releases. For us running it on every push is too costly (we don't pay for Github Actions). I will add a mention about that.

villelahdenvuo commented:
@jonabc I refactored the usage section based on your comments. I also removed duplication and separated the sections more while adding a full kitchen sink example. I also removed the scheduling example and I'll open another ticket for that.

jonabc commented on README.md:
```diff
@@ -72,38 +77,85 @@ on:
+jobs:
+  licensed:
+    steps:
+      - # ...
```
just making the comment a little more explicit 👍 

```suggestion
      - # environment setup ...
```

jonabc commented on README.md:
```diff
@@ -129,33 +181,35 @@ permissions:
+          cache: npm # cache dependencies for faster subsequent runs.
+      # install your projects dependencies
+      - run: npm install --production --ignore-scripts
+      - id: licensed
```
this looks like it needs a `jonabc/setup-licensed` step or similar right?

villelahdenvuo replied:
@jonabc When using the setup-ruby with the bundler-cache option it actually install all bundler dependencies automatically. But I can add back the configuration of the Gemfile `BUNDLE_GEMFILE: ${{ github.workspace }}/licensed.gemfile` to make it more clear.

jonabc commented on README.md:
````diff
@@ -72,38 +77,85 @@ on:
 ```
 
-Basic usage installing licensed gem using bundler + Gemfile
+### Full Node.js example
````
Do you think it makes sense to include this with the other examples?  Or alternatively is there a reason you put this after the authentication notes?

villelahdenvuo replied:
My idea was to put the full example at the bottom as it combines all the sections, including authentication. That way people can at least skim over the usage notes before they just copy and paste the full example.

Then if they get some authentication error they might remember seeing a header about authentication in the docs. 😄

jonabc commented:
the licensed failures are from this PR being a fork and not in any way related to the docs changes.  thanks for the updates 🙇
